### ISSUE TITLE
Resumable Uploads: POST for appending data?

### ISSUE THREAD

**Issue by Acconut** [httpwg/http-extensions#2962]
In https://mailarchive.ietf.org/arch/msg/httpbisa/m-iXTJQGdb8XpxGjZWxXGj6Siic/, Roy Fielding suggests POST to be used instead of PATCH + application/octet-stream for appending data to an upload resource:

> For example, use of the method PATCH for appending opaque data is incorrect. PATCH already has a defined semantic that could be used in parallel (e.g., another extension might use PATCH correctly). The resumable upload extension does not use a patch format, so both use of PATCH and creation of a new media type "application/partial-upload” are inappropriate.
>
> POST should be used for appending data to the upload resource. This matches the semantics of POST and is easier to deploy in practice. Each POST can use “application/octet-stream” (or not send any Content-Type, since it has no role to play in processing).

Since we define the upload resource in the first place, we can also specify the meaning of a POST request against it and don't have to use PATCH with a custom media type to define the message's semantics.

Personally, I think both methods (POST or PATCH) are fine. What do others think about this?

**Comment by LPardue** [comment]
I'd rather not do both.

Avoiding creating a media type has the positive of saving some effort. 

I tried to trace back the history of using PATCH in the first place and could not find a good answer. Does anyone have one?

**Comment by reschke** [comment]
On 26.11.2024 04:48, Guoye Zhang wrote:
> I believe that @reschke suggested using PATCH and defining a new media type initially.
>
> Personally I don't have a preference, either.

I'm not sure I did, but yes, that seems to be the right thing to do.

Best regards, Julian

**Comment by mnot** [comment]
Right, if we're going to use PATCH, the media type should describe how to handle the request content in a generic way. E.g., `application/append-these-bytes`.

That implies that it might be good to do so in a separate document, so it can be reused by other applications, rather than being perceived as specific to resumable uploads.

If that's thought of as too much work, POST is probably the right answer (with the semantics being defined by the specific resource type that's being baked into the resumable uploads spec).

But how could you resist creating a media type with a name like that?

**Comment by LPardue** [comment]
If its worth doing, it's worth doing right

**Comment by LPardue** [comment]
But also, didn't we have some of this discussion in the WG wrt https://datatracker.ietf.org/doc/draft-ietf-httpapi-patch-byterange/

**Comment by smatsson** [comment]
@LPardue We did. Both on the mailing list and on Github. 

https://github.com/httpwg/http-extensions/issues/2501 (byte range patch discussions)

https://mailarchive.ietf.org/arch/msg/httpbisa/UXTmMZaDNgtkMyA7broHeKL6vhg/ (contains discussions on POST vs PATCH)

https://mailarchive.ietf.org/arch/msg/httpbisa/aAfI1cZZDHnqhykFx4DoFClFocw/ (content type of PATCH and discussions if we can use it at all)

**Comment by Acconut** [comment]
> Right, if we're going to use PATCH, the media type should describe how to handle the request content in a generic way. E.g., `application/append-these-bytes`.
> 
> That implies that it might be good to do so in a separate document, so it can be reused by other applications, rather than being perceived as specific to resumable uploads.

The current draft (-05) intents to define the `application/partial-upload` media type (https://www.ietf.org/archive/id/draft-ietf-httpbis-resumable-upload-05.html#name-media-type-application-part) for use in PATCH requests when appending data, but this media type is specific to uploads and not very generic. We can consider renaming the media type and update its description to make it usable in other situations as well.

**Comment by mnot** [comment]
If it's spec-specific, that's good enough, I think.

**Comment by Acconut** [comment]
This topic was discussed during IETF 122. The feedback was that while POST and PATCH both fit semantically for appending to an upload resource, there is no strong reason to move away from the current PATCH + application/partial-upload media type.

Hence, I'll close this for now.

**Comment by GrantGryczan** [comment]
Should I open a new issue to express this concern?

**Comment by GrantGryczan** [comment]
I thought of that, but that seems strange to me. I guess technically that's fine and addresses my concern as an implementer, but it really seems contradictory to the semantics of `PATCH` being atomic in general. It's counterintuitive to define a `PATCH` media type that says "ignore the usual semantics of `PATCH` requests; I work differently", if that's even allowed.

**Comment by Acconut** [comment]
This exact wording of RFC 5789 was discussed at IETF 119 as well :) Please take a look at slide 7 of https://httpwg.org/wg-materials/ietf119/resumable-uploads.pdf, the corresponding minutes at https://httpwg.org/wg-materials/ietf119/minutes.html#resumable-uploads or alternatively the recording at https://meetecho-player.ietf.org/playout/?session=IETF119-HTTPBIS-20240319-0730.

The outcome was that the use of PATCH falls within the intentions of RFC 5789, which aim to ensure that a representation or resource is not left in an invalid state. This is the case for an upload resource after applying a PATCH request with a partially received body. The client is always able to continue the upload and interact with the upload resource.

Furthermore, RFC 5789 gives some room for when a PATCH is applied successfully:

> The determination of what constitutes a successful
   PATCH can vary depending on the patch document and the type of
   resource(s) being modified.

Hence we believe the use of PATCH is compatible with RFC 5789.

**Comment by GrantGryczan** [comment]
Thank you for the pointers. I believe that only addresses the last line of my comment:

> It could also be argued resumable uploads inherently violate this requirement already, because resumable uploads are certainly not atomic.

But I believe this fails to address my main concern. In fact, see this comment from the discussion:

> Martin Thomson: Agreed. Can go further and just say that any change in flight you will not see the partial change when querying the resource from elsewhere, which won’t violate the RFC.

This is exactly what I hope to avoid. I _want_ partial changes to be visible elsewhere, so that one client can stream a download of a resource while another client uploads it. Sure, you can define your partial upload media type such that it's considered "complete" to only apply part of it. But you can't define it such that each part of it being incrementally applied is its own atomic "completion" that can be observed from outside. Otherwise, it's inherently not atomic.

**Comment by Acconut** [comment]
The resumable uploads draft does not address in what form or way the uploaded representation is made available to others. The processing of the representation is controlled by the resource that was initially targeted and the semantics of the corresponding request.

**Comment by GrantGryczan** [comment]
Yes, but incrementally making parts of a single PATCH available to others while the patch document is processed seems to violate the PATCH spec, and that's something I think is important to be allowed to do. I'm not sure what your point is.

**Comment by guoye-zhang** [comment]
@GrantGryczan I think we should follow the spirit of the requirement rather than focusing on the exact wording. The atomicity requirement of PATCH is to prevent exposing intermediate states that might result in data corruption or expose security vulnerabilities. However, incrementally processing the request body during a resumable upload is no worse than processing a regular request body stream.

The requirement of PATCH is that the partially modified representation cannot be exposed outside. There is no requirement that we can't incrementally process the partially modified representation. Can we instead consider the entire series of PATCHes atomic for a resumable upload?

**Comment by GrantGryczan** [comment]
@guoye-zhang I wish the PATCH spec's wording matched its intent if that is the intent, because I'm not confident that it is. I wouldn't be surprised if it's also just an arbitrary semantic property of the PATCH method that is communicated to users of an HTTP application's API via that method, and that possibility makes me uncomfortable. After all, I don't believe any such requirement applies to POST, and this isn't just a recommendation but a requirement. But I will just violate the letter of the PATCH spec here, because while I'm not 100% sure atomicity isn't in the spirit of PATCH, I am 100% sure atomicity isn't in the spirit of resumable uploads. I just don't know why we don't choose an HTTP method that accurately reflects that.

> incrementally processing the request body during a resumable upload is no worse than processing a regular request body stream.

I think you're misunderstanding what I'm saying the problem is. I'm not talking about the regular processing inherent to resumable uploads. A particular application may concurrently expose a streaming download to another client from an in-progress upload (possibly an upload done in a single large PATCH request after an empty POST request), and that _is_ exposing the partially modified representation outside. Both each PATCH and the series of PATCHes would intentionally be measurably not atomic as a core feature of that application's functionality, even to a different client than the uploader.

**Comment by guoye-zhang** [comment]
@GrantGryczan This was the very first presentation about resumable uploads: https://datatracker.ietf.org/meeting/114/materials/slides-114-httpbis-resumable-uploads-01.pdf We did not use PATCH back then, but switched to PATCH based on the suggestions from the workgroup sessions and mailing list discussions.

POST is disliked by most people because it has very little defined semantics. It’s a catch all upload method that people use when taking a shortcut.

IMO PATCH can define the semantics of the resources being patched, but it cannot define the semantics of other resources. As long as we don’t serve the partial content from the temporary upload resource, it is OK to define a different resource or process the incremental content elsewhere. E.g. it can’t force a different process running on the same machine to not see the file actively being patched.

That being said, I personally don’t care much what method we use since it’s a trivial change as an implementer. If you think it’s an important enough issue to reopen and discuss during the workgroup session, we could.

**Comment by GrantGryczan** [comment]
I wasn't aware that many others find POST unfavorable. If the consensus is to prefer PATCH, then my objection is probably not important enough to warrant reconsideration (even though I personally stand by it). Nevertheless, I appreciate the time you spent on this discussion!
